File: src/php/Bridge.php
<?php
/**
 * PHP-Java-Bridge provides a TCP- or UDP-based connection between PHP (Client) and Java (Server)
 * 
 * licensed under the Apache License 2.0:
 * 
 * Permissions:
 *  - Commercial use
 *  - Modification
 *  - Distribution
 *  - Patent use
 *  - Private use
 * 
 * Limitiations:
 *  - Trademark use
 *  - Liability
 *  - Warranty
 * 
 * Conditions:
 *  - License and copyright notice
 *  - State changes
 * 
 * @link License: https://github.com/Synt4xErr0r4/PHP-Java-Bridge/blob/master/LICENSE
 * @link GitHub Repository: https://github.com/Synt4xErr0r4/PHP-Java-Bridge/
 * @link Wiki: https://github.com/Synt4xErr0r4/PHP-Java-Bridge/wiki
 * 
 * @version 1.0
 * @author SyntaxError404, 2020
 */
namespace phpjava;

use Exception;
use PhpAes\Aes;

require_once 'Packet.php';
require_once 'exceptions.php';
require_once 'lib/Aes.php';

define('BRIDGE_TCP',0);
define('BRIDGE_UDP',1);

/**
 * send a warning message if the 'E_WARNING' bitmask for 'error_reporting(?int)' is set
 */
function warn(string $msg) {
    if((E_WARNING&error_reporting())!=0)
        echo"<b>Warning:</b> $msg<br/>";
}
/**
 * send a notice message if the 'E_NOTICE' bitmask for 'error_reporting(?int)' is set
 */
function notice(string $msg) {
    if((E_NOTICE&error_reporting())!=0)
        echo"<b>Notice:</b> $msg<br/>";
}

/**
 * The connection between PHP and Java
 * 
 * Data can be transmitted via the Packet class
 * 
 * @see phpjava\Packet
 */
class Bridge {

    private$sock,$hostname,$port,$useAES,$passwd,$maxPacketSize,$method;

    /**
     * Instantiates a new PHP-Java-Bridge
     * 
     * @param method the method to be used: either TCP (BRIGE_TCP) or UDP (BRIDGE_UDP)
     * @param hostname the address (IPv4 or domain) to connect to the server
     * @param port the port of the server. default 8998
     * @param useAES whether or not the data should be AES-256-CBC encrypted. default false
     * @param passwd only required if useAES is true. Hashed with SHA3-256. default null
     * @param maxPacketSize the maximum size of a Packet. default 65535
     */
    public function __construct(int $method,string $hostname,int $port=8998,bool $useAES=false,?string $passwd=null,int $maxPacketSize=65535) {
        $this->hostname=$hostname;

        if($port<0||$port>65535)
            throw new Exception("Port out of range: $port: [0;65,535]");

        if(is_null($passwd)&&$useAES)
            throw new Exception("AES required a password, got null instead");

        $this->port=$port;
        $this->useAES=$useAES;
        $this->passwd=is_null($passwd)?null:hash('sha3-256',$passwd,true);

        if($maxPacketSize>0x7fffffff)
            throw new Exception("max. Packet size out of range: $maxPacketSize [6;2,147,483,647]");

        $this->maxPacketSize=$maxPacketSize;

        if($method!=BRIDGE_TCP&&$method!=BRIDGE_UDP)
            throw new Exception("Unrecognized method: $method");

        $this->method=$method;
    }

    /**
     * disconnects when destructed
     */
    public function __destruct() {
        if($this->sock==null)
            return;

        $this->disconnect();
        $this->sock=null;
    }

    /**
     * TCP: establishes connection to the server
     * UDP: initialized socket (doesn't throw Exception if the server is not reachable)
     */
    public function connect() {
        if(!is_null($this->sock))
            throw new ConnectionAlreadyEstablishedException("Already connected");

        if($this->method==BRIDGE_TCP)
            $this->sock=socket_create(AF_INET,SOCK_STREAM,SOL_TCP);
        else$this->sock=socket_create(AF_INET,SOCK_DGRAM,SOL_UDP);

        socket_connect($this->sock,$this->hostname,$this->port);

        if(!$this->sock) {
            $errno=socket_last_error();
            $errstr=socket_strerror($errno);
            throw new ConnectionFailedException("Couldn't connect to server: $errstr [#$errno]");
        }
    }
    public function disconnect() {
        if(is_null($this->sock))
            throw new ConnectionNotEstablishedYetException("Not connected yet");

        if($this->method==BRIDGE_TCP)
            socket_close($this->sock);
        $this->sock=null;
    }

    /**
     * Sends and receives a packet.
     * 
     * @param packet the Packet to be sent to the server
     * 
     * @return Packet the Packet received from the server
     */
    public function sendPacket(Packet $packet):Packet {
        
        if(is_null($this->sock))
            throw new ConnectionNotEstablishedYetException("Not connected yet");

        $raw=$packet->raw();

        $data=$this->encrypt($raw);
        $len=strlen($data);

        if($len+6>$this->maxPacketSize)
            throw new Exception("Packets exceeds max. allowed size: $len");

        $message=pack('C',$packet->getPacketID()).pack('C',isLittleEndian()?1:0).pack('N',$len&0x7fffffff).$data;

        $packet->__destruct();
        
        if(!socket_sendto($this->sock,$message,strlen($message),0,$this->hostname,$this->port)) {
            $errno=socket_last_error();
            $errstr=socket_strerror($errno);
            throw new Exception("Couldn't send packet: $errstr [#$errno]");
        }
        
        if(($len=socket_recv($this->sock,$buffer,$this->maxPacketSize,$this->method==BRIDGE_UDP?0:MSG_WAITALL))===FALSE) {
            $errno=socket_last_error();
            $errstr=socket_strerror($errno);
            throw new Exception("Couldn't send packet: $errstr [#$errno]");
        }

        $pid=unpack('C',$buffer)[1];
        $size=unpack('N',substr($buffer,2,4))[1];

        if($len<$size+6)
            throw new Exception("Received too few bytes: Expected at least ".($size+6).", got $len instead");

        $response=new Packet($pid);
        $response->setAndValidate($this->decrypt(substr($buffer,6,$size)));

        return$response;
    }

    /**
     * @param plainText the text to be encrypted
     * 
     * @return string the AES-256-CBC encrypted string. The password is hashed with SHA3-256
     */
    private function encrypt(string $plainText):string {
        if(!$this->useAES)
            return base64_encode(pack('N',strlen($plainText)).$plainText);

        $iv=openssl_random_pseudo_bytes(16);

        if(!$iv) {
            $err=error_get_last()or['type'=>0,'message'=>'null'];
            throw new Exception("Couldn't generate IV: ".$err['message']." [#".$err['type']."]");
        }

        $aes=new Aes($this->passwd,'CBC',$iv);
        
        return base64_encode($iv.$aes->encrypt(pack('N',strlen($plainText)).$plainText));
        //return base64_encode($iv.openssl_encrypt(pack('N',strlen($plainText)).$plainText,'AES-256-CBC',$this->passwd,OPENSSL_RAW_DATA,$iv));
    }
    /**
     * @param cipherText the encrypted text
     * 
     * @return string the decrypted plain text
     */
    private function decrypt(string $cipherText):string {
        echo"<hr>decrin ".strlen($cipherText)." => ";
        foreach(array_map(fn($x)=>unpack('c',pack('C',ord($x)))[1],str_split($cipherText))as$v)echo"$v, ";
        echo"<hr>";

        $cipherText=base64_decode($cipherText);

        if(!$this->useAES)
            return substr($cipherText,4,unpack('N',$cipherText));
        
        $iv=substr($cipherText,0,16);
        $cipherText=substr($cipherText,16);

        $aes=new Aes($this->passwd,'CBC',$iv);

        $buf=$aes->decrypt($cipherText);
        //$buf=openssl_decrypt($cipherText,'AES-256-CBC',$this->passwd,OPENSSL_RAW_DATA,$iv);
        
        echo"<hr>decr ".strlen($buf)." => ";
        foreach(array_map(fn($x)=>unpack('c',pack('C',ord($x)))[1],str_split($buf))as$v)echo"$v, ";
        echo"<hr>";

        return substr($buf,4,unpack('N',$buf)[1]);
    }

}

?>

File: src/java/api/syntaxerror/phpjavabridge/Bridge.java
package api.syntaxerror.phpjavabridge;

import java.lang.Thread.UncaughtExceptionHandler;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.util.Arrays;
import java.util.Base64;
import java.util.HashMap;
import java.util.Map;

import javax.crypto.Cipher;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;

/**
 * PHP-Java-Bridge provides a TCP- or UDP-based connection between PHP (Client) and Java (Server)<br>
 * <br>
 * licensed under the Apache License 2.0:<br>
 * <br>
 * Permissions:
 * <ul>
 *  <li>Commercial use</li>
 *  <li>Modification</li>
 *  <li>Distribution</li>
 *  <li>Patent use</li>
 *  <li>Private use</li>
 * </ul>
 * 
 * Limitiations:
 * <ul>
 *  <li>Trademark use</li>
 *  <li>Liability</li>
 *  <li>Warranty</li>
 * </ul>
 * 
 * Conditions:
 * <ul>
 *  <li>License and copyright notice</li>
 *  <li>State changes</li>
 * </ul>
 * 
 * License: <a href=https://github.com/Synt4xErr0r4/PHP-Java-Bridge/blob/master/LICENSE>https://github.com/Synt4xErr0r4/PHP-Java-Bridge/blob/master/LICENSE</a><br>
 * GitHub Repository: <a href=https://github.com/Synt4xErr0r4/PHP-Java-Bridge/>https://github.com/Synt4xErr0r4/PHP-Java-Bridge/</a><br>
 * Wiki: <a href=https://github.com/Synt4xErr0r4/PHP-Java-Bridge/wiki>https://github.com/Synt4xErr0r4/PHP-Java-Bridge/wiki</a><br>
 * 
 * @version 1.0
 * @author SyntaxError404, 2020
 */
public abstract class Bridge {

	protected int port,maxPacketLength;
	private boolean useAES;
	private byte[]password;
	protected Thread thread;
	protected Map<Integer,PacketHandler>handlers;
	protected UncaughtExceptionHandler exceptionHandler;
	
	/**@see {@link BridgeImpl.TCP#TCP(int, boolean, String, int)}
	 * @see {@link BridgeImpl.UDP#UDP(int, boolean, String, int)}
	 */
	Bridge(int port,boolean useAES,String password,int maxPacketLength) {
		handlers=new HashMap<>();
		
		this.port=port;
		
		if(useAES) {
			if(password==null)
				throw new NullPointerException("AES requires a password, got null instead");
			
			this.useAES=true;
			
			try {
				MessageDigest md=MessageDigest.getInstance("SHA3-256");
				this.password=md.digest(password.getBytes(StandardCharsets.UTF_8));
			} catch(NoSuchAlgorithmException e) {
				throw new RuntimeException("Couldn't find SHA3-256 algorithm"); 
			}
			
		}
		
		if(maxPacketLength<6)
			throw new IllegalArgumentException("maxPacketLength must be greater than or equal to 6"); 
		
		this.maxPacketLength=maxPacketLength;
		
		thread=new Thread(()->{});
	}
	
	public int getPort() {
		return port;
	}

	/**
	 * Starts the Server-Thread.
	 */
	public void start() {
		thread.start();
	}

	/**
	 * Stops the Server Thread.<br>
	 * You cannot call {@link #start()} afterwards.
	 * 
	 * @throws InterruptedException thrown by {@link Thread#join()}
	 */
	public void stop()throws InterruptedException {
		thread.join();
	}

	/**
	 * If a {@link Packet} couldn't be handled (because there was no {@link PacketHandler} set via {@link #setPacketHandler(PacketHandler, int...)}), this handler is called.<br>
	 * <br>
	 * If there is no default {@link PacketHandler} either, a {@link UnsupportedOperationException} is thrown.
	 * 
	 * @param handler the new default {@link PacketHandler}
	 */
	public void setDefaultPacketHandler(PacketHandler handler) {
		handlers.put(-1,handler);
	}

	/**
	 * Defines a {@link PacketHandler} for specific Packet-IDs.<br>
	 * <br>
	 * If there is no {@link PacketHandler} defined for a Packet-ID and no default {@link Packet} is set via {@link #setDefaultPacketHandler(PacketHandler)}, a
	 * {@link UnsupportedOperationException} is thrown if such a {@link Packet} is received
	 * 
	 * @param handler the {@link PacketHandler}
	 * @param pids the Packet-IDs where this {@link PacketHandler} should be used
	 */
	public void setPacketHandler(PacketHandler handler,int...pids) {
		for(int pid:pids)
			if(pid<0||pid>255)
				throw new IndexOutOfBoundsException("Invalid Packet-ID: " +pid);
			else handlers.put(pid,handler);
	}
	
	/**
	 * A {@link UncaughtExceptionHandler} prevents the {@link Bridge} from stopping if an Exception occured. 
	 * 
	 * @param exceptionHandler the {@link UncaughtExceptionHandler}
	 */
	public void setExceptionHandler(UncaughtExceptionHandler exceptionHandler) {
		this.exceptionHandler=exceptionHandler;
	}
	
	/**
	 * internal use only<br><br>
	 * 
	 * AES-256-CBC encryption with SHA3-256 hashed password
	 * 
	 * @param plainText the raw data
	 * 
	 * @return the encrypted data
	 */
	public final byte[]encrypt(byte[]plainText) {
		int len=plainText.length;
		
		byte[]raw=new byte[4+len+16-((4+len)%16)];
		
		raw[0]=(byte)((len>>>24)&0xFF);
		raw[1]=(byte)((len>>>16)&0xFF);
		raw[2]=(byte)((len>>>8)&0xFF);
		raw[3]=(byte)(len&0xFF);
		
		System.arraycopy(plainText,0,raw,4,len);
		
		
		if(!useAES)
			return Base64.getEncoder().encode(raw);
		
		byte[]iv=new byte[16],encrypted;
		new SecureRandom().nextBytes(iv);
		
		try {
			Cipher cipher=Cipher.getInstance("AES/CBC/NoPadding");
			cipher.init(Cipher.ENCRYPT_MODE,new SecretKeySpec(password,"AES"),new IvParameterSpec(iv));
			encrypted=cipher.doFinal(raw);
		} catch(Exception e) {
			throw new RuntimeException(e);
		}
		
		byte[]full=new byte[16+encrypted.length];
		
		System.arraycopy(iv,0,full,0,16);
		System.arraycopy(encrypted,0,full,16,encrypted.length);
		
		return Base64.getEncoder().encode(full);
	}
	/**
	 * internal use only<br><br>
	 * 
	 * AES-256-CBC decryption with SHA3-256 hashed password
	 * 
	 * @param cipherText the encrypted data
	 * 
	 * @return the decrypted plain text
	 */
	public final byte[]decrypt(byte[]cipherText) {
		cipherText=Base64.getDecoder().decode(cipherText);
		
		if(!useAES) {
			int len=((cipherText[0]&0xFF)<<24)|
					((cipherText[1]&0xFF)<<16)|
					((cipherText[2]&0xFF)<<8)|
					(cipherText[3]&0xFF);
			
			return Arrays.copyOfRange(cipherText,4,4+len);
		}
		
		byte[]iv=Arrays.copyOf(cipherText,16);
		cipherText=Arrays.copyOfRange(cipherText,16,cipherText.length);
		
		try {
			Cipher cipher=Cipher.getInstance("AES/CBC/NoPadding");
			cipher.init(Cipher.DECRYPT_MODE,new SecretKeySpec(password,"AES"),new IvParameterSpec(iv));
			byte[]buf=cipher.doFinal(cipherText);
			int len=((buf[0]&0xFF)<<24)|
					((buf[1]&0xFF)<<16)|
					((buf[2]&0xFF)<<8)|
					(buf[3]&0xFF);
			
			return Arrays.copyOfRange(buf,4,4+len);
		} catch(Exception e) {
			throw new RuntimeException(e);
		}
	}
	
	// STATIC METHODS

	/**
	 * Creates a new {@link BridgeImpl.TCP TCP-Bridge}<br><br>
	 * 
	 * see {@link Bridge#Bridge(int, boolean, String, int)}<br>
	 * see {@link BridgeImpl.TCP#TCP(int, boolean, String, int)}
	 * 
	 * @return TCP-based {@link Bridge}
	 */
	public static Bridge newTCP() {
		return newTCP(8998,false,null,65535);
	}
	/**
	 * Creates a new {@link BridgeImpl.TCP TCP-Bridge}<br><br>
	 * 
	 * see {@link Bridge#Bridge(int, boolean, String, int)}<br>
	 * see {@link BridgeImpl.TCP#TCP(int, boolean, String, int)}
	 * 
	 * @param port the port to be used
	 * 
	 * @return TCP-based {@link Bridge} 
	 */
	public static Bridge newTCP(int port) {
		return newTCP(port,false,null,65535);
	}
	/**
	 * Creates a new {@link BridgeImpl.TCP TCP-Bridge}<br><br>
	 * 
	 * see {@link Bridge#Bridge(int, boolean, String, int)}<br>
	 * see {@link BridgeImpl.TCP#TCP(int, boolean, String, int)}
	 * 
	 * @param port the port to be used
	 * @param maxPacketLength the max. length of a single {@link Packet}
	 * 
	 * @return TCP-based {@link Bridge}
	 */
	public static Bridge newTCP(int port,int maxPacketLength) {
		return newTCP(port,false,null,maxPacketLength);
	}
	/**
	 * Creates a new {@link BridgeImpl.TCP TCP-Bridge}<br><br>
	 * 
	 * see {@link Bridge#Bridge(int, boolean, String, int)}<br>
	 * see {@link BridgeImpl.TCP#TCP(int, boolean, String, int)}
	 * 
	 * @param port the port to be used
	 * @param useAES whether or not AES-256-CBC should be used. requires {@code password}
	 * @param password the password to be used. Required by AES
	 * 
	 * @return TCP-based {@link Bridge}
	 */
	public static Bridge newTCP(int port,boolean useAES,String password) {
		return newTCP(port,useAES,password,65535);
	}
	/**
	 * Creates a new {@link BridgeImpl.TCP TCP-Bridge}<br><br>
	 * 
	 * see {@link Bridge#Bridge(int, boolean, String, int)}<br>
	 * see {@link BridgeImpl.TCP#TCP(int, boolean, String, int)}
	 * 
	 * @param port the port to be used
	 * @param useAES whether or not AES-256-CBC should be used. requires {@code password}
	 * @param password the password to be used. Required by AES
	 * @param maxPacketLength the max. length of a single {@link Packet}
	 * 
	 * @return TCP-based {@link Bridge}
	 */
	public static Bridge newTCP(int port,boolean useAES,String password,int maxPacketLength) {
		return new BridgeImpl.TCP(port,useAES,password,maxPacketLength);
	}

	/**
	 * Creates a new {@link BridgeImpl.UDP UDP-Bridge}<br><br>
	 * 
	 * see {@link Bridge#Bridge(int, boolean, String, int)}<br>
	 * see {@link BridgeImpl.UDP#UDP(int, boolean, String, int)}
	 * 
	 * @return UDP-based {@link Bridge}
	 */
	public static Bridge newUDP() {
		return newUDP(8998,false,null,65535);
	}
	/**
	 * Creates a new {@link BridgeImpl.UDP UDP-Bridge}<br><br>
	 * 
	 * see {@link Bridge#Bridge(int, boolean, String, int)}<br>
	 * see {@link BridgeImpl.UDP#UDP(int, boolean, String, int)}
	 * 
	 * @param port the port to be used
	 * 
	 * @return UDP-based {@link Bridge} 
	 */
	public static Bridge newUDP(int port) {
		return newUDP(port,false,null,65535);
	}
	/**
	 * Creates a new {@link BridgeImpl.UDP UDP-Bridge}<br><br>
	 * 
	 * see {@link Bridge#Bridge(int, boolean, String, int)}<br>
	 * see {@link BridgeImpl.UDP#UDP(int, boolean, String, int)}
	 * 
	 * @param port the port to be used
	 * @param maxPacketLength the max. length of a single {@link Packet}
	 * 
	 * @return UDP-based {@link Bridge}
	 */
	public static Bridge newUDP(int port,int maxPacketLength) {
		return newUDP(port,false,null,maxPacketLength);
	}
	/**
	 * Creates a new {@link BridgeImpl.UDP UDP-Bridge}<br><br>
	 * 
	 * see {@link Bridge#Bridge(int, boolean, String, int)}<br>
	 * see {@link BridgeImpl.UDP#UDP(int, boolean, String, int)}
	 * 
	 * @param port the port to be used
	 * @param useAES whether or not AES-256-CBC should be used. requires {@code password}
	 * @param password the password to be used. Required by AES
	 * 
	 * @return UDP-based {@link Bridge}
	 */
	public static Bridge newUDP(int port,boolean useAES,String password) {
		return newUDP(port,useAES,password,65535);
	}
	/**
	 * Creates a new {@link BridgeImpl.UDP UDP-Bridge}<br><br>
	 * 
	 * see {@link Bridge#Bridge(int, boolean, String, int)}<br>
	 * see {@link BridgeImpl.UDP#UDP(int, boolean, String, int)}
	 * 
	 * @param port the port to be used
	 * @param useAES whether or not AES-256-CBC should be used. requires {@code password}
	 * @param password the password to be used. Required by AES
	 * @param maxPacketLength the max. length of a single {@link Packet}
	 * 
	 * @return UDP-based {@link Bridge}
	 */
	public static Bridge newUDP(int port,boolean useAES,String password,int maxPacketLength) {
		return new BridgeImpl.UDP(port,useAES,password,maxPacketLength);
	}
	
}
